Introduce `moveBefore()` state-preserving atomic move API

domfarolino opened the pull request:
This PR introduces a new DOM API on the `ParentNode` mixin: `moveBefore()`. It mirrors `insertBefore()` in shape, but defers to a new DOM manipulation primitive that this PR adds in service of this new API: the "move" primitive. The move primitive contains some of the DOM tree bookkeeping steps from the remove primitive, as well as the insert primitive, and does three more interesting things:
 1. Calls the [moving steps hook](https://whatpr.org/dom/1307.html#concept-moving-steps-ext) with the moved node and the old parent (possibly null, just like the removing steps)
 2. Queues a custom element callback reaction for the `connectedMoveCallback()`
 3. Queues two back-to-back mutation record tasks: one for removal from the old parent; one for insertion into the new one

The power of the move primitive comes from the fact that the algorithm does not defer to the traditional insert and removal primitives, and therefore does not invoke the [removing steps](https://dom.spec.whatwg.org/#concept-node-remove-ext) and [insertion steps](https://dom.spec.whatwg.org/#concept-node-insert-ext). This allows most state to be preserved by default (i.e., we don't tear down iframes, or close dialogs). Sometimes, the insertion/removing step overrides in other specifications have steps that do need to be performed during a move anyways. These specifications are expected to override the moving steps hook and perform the necessary work accordingly. See https://github.com/whatwg/html/pull/10657 for HTML.

Closes #1255. Closes #1335.

Remaining tasks (some will be PRs in other standards):

- [x] Custom element integration
- [x] Keep popovers open
- [x] Don't call post-connection steps if state-preserving atomic move is in progress
- [x] Don't call becomes connected / becomes browsing-context
- [x] Only disconnect subframes on removal when state-preserving atomic move is **not** in progress
- [x] Keep dialogs open: see [removing steps](https://html.spec.whatwg.org/C#the-dialog-element:html-element-removing-steps)
- [x] img/source: this shouldn't count as a [relevant mutation](https://html.spec.whatwg.org/C#relevant-mutations)
- [x] Preserve fullscreen <!--analogous to https://crrev.com/c/5466094-->
- [x] Preserve focus <!--analogous to https://crrev.com/c/5465724-->
    - Need to resolve `focusin` event semantics
- ~[ ] Don't reset animations / transitions. See [here](https://drafts.csswg.org/css-transitions/#starting) <!--analogous to https://crrev.com/c/5458120)-->
    - Maybe nothing needs to be done here. Given [how element removals are handled](https://drafts.csswg.org/css-transitions/#transition-cancel:~:text=If%20an%20element%20is%20no%20longer%20in%20the%20document%2C%20implementations%20must%20cancel%20any%20running%20transitions%20on%20it%20and%20remove%20transitions%20on%20it%20from%20the%20completed%20transitions), the spec does NOT require transitions to be removed from the UA's set of _running transitions_ for moved nodes since they are never removed from the Document.~
- ~[ ] Preserve text-selection. See [set the selection range](https://html.spec.whatwg.org/C#set-the-selection-range). Edit: Nothing needs to be done here. Selection metadata (i.e., `selectionStart` and kin) is preserved by default in browsers, consistent with HTML (no action is taken on removal). The UI behavior of the selection not being highlighted is a side-effect of the element losing focus~
- [ ] Selection API: don't reset the Document's [selection](https://w3c.github.io/selection-api/#dfn-selection)
    - Updates to the selection range should happen according to how the DOM Standard primitives update ranges. The Selection API specification admits as much, by [deferring to the](https://w3c.github.io/selection-api/#responding-to-dom-mutations) insert and removal algorithms. Therefore, we should reference the move primitive from the Selection API specification, and ensure that the move primitive in _this DOM Standard PR_ updates live ranges correctly: https://github.com/w3c/selection-api/pull/341
    - `selectionchange` event: We've decided to allow `selectionchange` event to still fire, [since it is queued in a task](https://w3c.github.io/selection-api/#ref-for-dfn-schedule-a-selectionchange-event-1). No changes for this part are required.
- [x] Pointer event state reset: see [here](https://w3c.github.io/pointerevents/#suppressing-a-pointer-event-stream)
    - The specification does not reset an element's previously-set pointer capture upon DOM node insertion or removal. While this seems like a gap in that specification (https://github.com/w3c/pointerevents/issues/526), it appears that nothing needs to be done in that specification to satisfy this preservation semantic.
    - Test: https://github.com/web-platform-tests/wpt/blob/master/dom/nodes/moveBefore/tentative/pointer-events.html
- [x] Hide input/select picker: [here](https://html.spec.whatwg.org/C#show-the-picker%2C-if-applicable)
    - Nothing in the specification seems to hook into DOM for removal when the picker dialog is shown, so no work needs to be done in this PR. See https://github.com/whatwg/html/issues/10743 for a follow-up.
- [x] Preserve pointer lock: [here](https://w3c.github.io/pointerlock/#requestPointerLock)
    - When a ["pointer-lock target becomes disconnected"](https://w3c.github.io/pointerlock/#ref-for-dfn-pointer-lock-target-12), pointer lock is exited. Since the "becomes disconnected" hook is not run during an atomic move, pointer lock is preserved naturally.
- [ ] Containment: keep last remembered size(see [here](https://drafts.csswg.org/css-sizing-4/#last-remembered))

<!-- Template: -->

- [x] At least two implementers are interested (and none opposed):
   * Chromium: https://issues.chromium.org/issues/40150299
   * WebKit: https://github.com/WebKit/standards-positions/issues/375
   * Gecko: https://github.com/mozilla/standards-positions/issues/1053
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/tree/master/dom/nodes/moveBefore/tentative (need to mark as non-tentative)
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://issues.chromium.org/issues/40150299
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1923880
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=281223
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/mdn/issues/613 & the `impacts-documentation` label
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/dom/1307.html" title="Last updated on Feb 27, 2025, 4:30 PM UTC (bfbcfd9)">Preview</a> | <a href="https://whatpr.org/dom/1307/b64559c...bfbcfd9.html" title="Last updated on Feb 27, 2025, 4:30 PM UTC (bfbcfd9)">Diff</a>

domfarolino commented on dom.bs:
```diff
@@ -4986,6 +5033,27 @@ within <a>this</a>.
+ <li><p>If <var>perform state-preserving atomic move</var> is true, then set <a>this</a>'s <a
+ for=Node>node document</a>'s <a>state-preserving atomic move in progress</a> to true.</p></li>
+
+ <li><p>Let <var>return node</var> be the result of <a>pre-inserting</a> <var>node</var> into
+ <a>this</a> before <var>child</var>.</p></li>
```
Yeah I think initially I saw it as a separate primitive too, but the more I looked at it, the more the difference between the two seemed really subtle. It mostly has to do with MutationObservers (and half of the relevant logic here is tucked away in the "remove" primitive) and *not* running the post-connection steps. So I feel like we'd end up with a near line-by-line copy of "insert", modulo one or two small differences. I'll take another look to see if my intuition is accurate, but I do kinda suspect this is where we'd end up.

annevk replied:
It'd be good to have more details. Because you can't move a `DocumentFragment` so a vast number of checks don't apply. Adopt won't ever run.

There might be a number of range and shadow tree checks that end we end up duplicating, but perhaps that calls for abstracting those instead. At least to me a state flag seems rather unappealing.

domfarolino replied:
It's possible there is enough from "insert" that we'd omit in "move", to make it creation worth it, sure.

> At least to me a state flag seems rather unappealing.

With a separate "move" primitive _and_ the decision to throw when we can't "move", we could get rid of the state variable that's currently in "insert" (i.e., `<var>statePreservingAtomicMoveInProgress</var>`). But the state bool on _`Document`_ is what other specifications will refer to in their removal steps to react to a move appropriately, so I'm not sure we can get rid of _that_ one.

annevk replied:
Hmm, wouldn't we give other specifications "move steps"?

domfarolino replied:
I thought about this a bit more. I think we're happy to create a new "move" primitive algorithm following the "insert" and "remove" algorithms. As you mentioned, it will be a simpler algorithm than "insert" since we'll avoid the `DocumentFragment` stuff — oh and also the post-connection stuff, among maybe more things.

There's still the discussion around introducing either (**a**) new "move" steps/hook that other specs provide to react to moves, vs (**b**) a Document-scoped boolean that the new "move" algorithm sets, for existing insertion/removal steps to use, to not reset certain state inside those steps.

Option (b) is the direction of the current PR, which we think makes sense after implementing this in Chromium. I think this is because so far we've found no behavior that we specifically want to do during a move that isn't done by insertion/removal — so any "move steps" we provided would just be empty. **Does that make sense?**

Now, I think the real question here is whether we want to:

1. Keep running the [insertion](https://dom.spec.whatwg.org/#concept-node-insert-ext)/[removal](https://dom.spec.whatwg.org/#concept-node-remove-ext) steps during a "move" operation, and just pepper `if during atomic move is true, don't run some of these substeps!` wherever state would ordinarily be reset
    - Cancel out [the iframe removal steps](https://html.spec.whatwg.org/C#the-iframe-element:html-element-insertion-steps)
    - Cancel [img relevant mutations](https://html.spec.whatwg.org/C#reacting-to-dom-mutations:html-element-insertion-steps), to avoid re-firing load events
1. Simply not run insertion/removal hooks _at all_ during a move. This is only feasible if we determine that there are absolutely no steps in any spec's existing insertion/removal hooks that need to run during a move. We'll have to do an audit to determine this, which I'm starting now.

So the way I see it, those are probably our options, as opposed to creating new move steps which I think from our findings so far, would be unused. (We'll course-correct if this finding changes).

smaug---- commented:
> * old target
> 
> * target
> 
> * moved node (I'm not sure you can ever move multiple at this point, but maybe we should allow for it in the mutation record design?)

Shouldn't the target node be all the time the same, it is just the siblings which change.
So we'd need only oldPreviousSibling and oldNextSibling. Oh, hmm, this isn't only about moving children but moving anything.

If this is really just remove and add back elsewhere, we could just reuse the existing childList MutationRecords, one for remove, one for adding node back, and possibly just add a flag to MutationRecord that it was about move.


(movedNodes is a bit confusing, since it seems to depend on the connectedness of the relevant nodes and it is apparently empty for the removal part. And it is unclear to me why we need the connectedness check. This is about basic DOM tree operations, and I'd assume those to work the same way whether or not the node is connected)

annevk commented:
Creating two separate mutation records that a consumer would have to merge to (fully) understand it's a move seems suboptimal?

I agree that it should probably work for disconnected nodes as well, but I don't think we want to support a case where the shadow-including root changes.

ajklein commented:
It's been a long time since I've thought about this stuff, but I'm inclined to agree with @smaug---- that creating a new type of `MutationRecord` feels unnecessary. Users of `MutationObserver` already have to do coalescing if they want to make sense of the stream of changes they observe. There are already other move-like operations, such as appending a child that's already somewhere else in the tree, which today generates two records.

domfarolino commented on dom.bs:
```diff
@@ -4986,6 +5033,27 @@ within <a>this</a>.
+ <li><p>Let <var>perform state-preserving atomic move</var> be true if <a>this</a> is
+ <a>connected</a>, <var>node</var> is <a>connected</a>, <a>this</a>'s <a for=Node>node
+ document</a> is <a>fully active</a>, and <a>this</a>'s <a for=/>root</a> is the same as
+ <var>node</var>'s <a for=/>root</a>.</p></li>.
```
Done!

smaug---- commented on dom.bs:
```diff
@@ -2652,6 +2652,50 @@ of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run the
+
+<ol>
+ <li><p>If either <var>parent</var> or <var>node</var> are not <a>connected</a>, then
+ <a>throw</a> a "{{HierarchyRequestError!!exception}}" {{DOMException}}.</p></li>
```
Hmm, why do we have this limitation. Why not allow move within the same tree? Basically why the condition about shadow-including root isn't enough?

domfarolino replied:
I believe we discussed this at TPAC. The idea is that `moveBefore()` is used to preserve state, so in cases where state cannot be preserved, we should not just fall back to usual `insertBefore()`-style behavior. Instead, we'll throw and let the developer handle this case themselves. Basically we let `moveBefore()` succeed only when we're sure we can perform a state-preserving move, and when disconnected destinations or origins are involved, the state cannot be preserved.

smaug---- replied:
I would definitely expect a move operation to succeed if one stays within a same tree, whether or not your connected to document.

annevk replied:
I don't see how connected <> disconnected would work. It would need a completely different code path, no?

I can see connected <> connected and disconnected <> disconnected. Someone (either Noam or Dominic) argued against disconnected <> disconnected earlier, but it seems reasonable to allow as well.

noamr replied:
I'm ok with both. One can say that disconnected<>disconnected doesn't have any state to preserve, so whether we throw or do the same as insertBefore is a judgement call.

annevk replied:
We should not do the same as insertBefore. It should run the moving steps (whether anyone cares about those for disconnected is a different question).

noamr replied:
Sure, I mean that it should be observably the same in the disconnected<>disconnected case.

domfarolino replied:
I've made this PR support both the `connected->connected` and `disconnected->disconnected` case, as of https://github.com/whatwg/dom/pull/1307/commits/2bbc8346167310cc3e7f8d2ea65e5151c9cc219f. I think that's sufficient, and we've heard that developers want this too, so I think that's enough to resolve things here.

tjallingt replied:
Disconnected -> connected could theoretically also been seen as stated preserving since the disconnected node didn't have any previous state so there is nothing to preserve.

Connected -> disconnected moves could destroy the state though so should definitely not be allowed.

domfarolino replied:
> Disconnected -> connected could theoretically also been seen as stated preserving since the disconnected node didn't have any previous state so there is nothing to preserve.

I think I disagree. The state could be null `contentDocument`, lack of focus, no form owner, and more etc. That "no state" is still a kind of state, and it'd either be:
 - **Changed** by becoming connected into a document, and running the many hooks that the platform does during that milestone — some of which initialize iframe setup, run script, etc. In this case, move just becomes insertion, which we don't want to do because the two are distinct primitives.
 - **Not changed**, by just inserting the element into the DOM and leaving it kind of dead/void. This would be completely new and I think kind of weird. For example `parent.moveBefore(disconnectedIframe, null)` would make `disconnectedIframe` finally connected, but with a null content document because we never initialized the setup. In that case, it feels like there was no reason to *move* it into the DOM in the first place. Same goes for other things with form owners, executable `<script>`s, etc. I'm not sold that this is a real use case.

smaug---- commented on dom.bs:
```diff
@@ -2652,6 +2652,50 @@ of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run the
+ "{{HierarchyRequestError!!exception}}" {{DOMException}}.</p></li>
+
+ <li><p>If <var>parent</var> is not an {{Element}} <a for=/>node</a>, then <a>throw</a> a
+ "{{HierarchyRequestError!!exception}}" {{DOMException}}.
```
Why moves in DocumentFragments aren't possible? Especially, what if parent is a ShadowRoot?

domfarolino replied:
DocumentFragments are always disconnected, right? We require that the moved node, and its new parent, are both connected to the DOM.

smaug---- replied:
Shadowroot is a DocumentFragment. I would expect moves under ShadowRoot being possible.
(I wasn't thinking the connectedness case here, only that one should be able reorder the children of a ShadowRoot)

domfarolino replied:
I didn't realize `ShadowRoot` was a `DocumentFragment`! Thanks for pointing this out. OK I've changed the spec to enable this (we can discuss connected-ness separately), and I've added a test for this in https://crrev.com/c/6012161. I'll resolve this now.

domfarolino commented on dom.bs:
```diff
@@ -2845,6 +2889,125 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+  need to update <a>live range</a> state when their <a for=range>start node</a> or
+  <a for=range>end node</a> is an <a>inclusive descendant</a> of the <var>node</var>. This is
+  because said <a>nodes</a> do not get removed from their <a>tree</a>, so ranges associated with
+  them stay intact.</p>
```
I don't think so. It should be equivalent to when a range's start container is set (to the first time) to a node after the end container as well. Here are tests for this, which include `intersectsNode` assertions to make sure the Range's state is not peculiar: https://github.com/web-platform-tests/wpt/blob/master/dom/nodes/moveBefore/tentative/live-range-updates.html.

smaug---- replied:
Ok, so something else still updates Range boundary points so that start is before end point? The note is then a bit confusing.

domfarolino replied:
> Ok, so something else still updates Range boundary points so that start is before end point?

No, a Range's start does not have to be before its end, in DOM tree order. Just like if you create a new Range whose start comes after its end in DOM tree order, the start's [position](https://dom.spec.whatwg.org/#concept-range-bp-position) will be "after" its end, which is observable via `intersectsNode()` since nodes that are after end but before start, will no longer `intersectsNode()` the Range. The same goes for when `moveBefore()` moves things around such that the range's start is after its end, in tree order.

smaug---- replied:
Wait, how do you create a Range which has its start after end? setStart/End etc will modify the other boundary point if the order isn't right.

domfarolino replied:
Ooof, OK then the spec indeed does leave ranges in a bad state, since what this PR allows (wrt `Range` + `moveBefore()`) is not possible with `Range` today. Great catch, @smaug----. In that case, I think the best way to deal with this would be to, during `moveBefore()`, detect if the moved node is the inclusive ancestor of a start or end boundary point of any range, and if that boundary point crosses its corresponding start/end, and collapse the node at the other boundary point. How does that sound to you?

smaug---- replied:
That wouldn't be consistent with other tree operations. I think I'd prefer a model where ranges were treated as if a node was removed and then added back to a different place in tree. 
(Kind of like how MutationRecords are handled)

domfarolino replied:
In that case, we wouldn't be able to preserve selection with `moveBefore()`, which is one of the pieces of state we were really hoping to preserve with this API. This is especially annoying when you `moveBefore()` the selection range's start container to somewhere earlier in the doc, or the end container somewhere later. The range is still totally valid, so it seems like we should be able to preserve selection. What do you think?

domfarolino replied:
Another option for partial preservation (discussed with @noamr) would be to preserve the selection if and only if the moved node is an inclusive ancestor of _both_ the start and end containers of the range. That way if the selection (or other range) is entirely contained within the moved node, we don't lose it.

noamr replied:
@smaug---- perhaps a useful middle ground can be that we preserve the selection only if the moved node is an ancestor of the entire range. This way if the range is moved while "intact", the selection would be preserved, but if parts of it are moved it's considered "broken" and it's dissolved. This is useful when moving whole elements which might have bits of text selected inside them.

smaug---- replied:
Not sure what  you mean with "wouldn't be able to preserve selection". Using the normal removal/addition rules  would preserve selection the usual way. The range just gets modified.

smaug---- replied:
Container approach (not modifying the range if full contained) would be still problematic if the container was moved from light dom to shadow dom, and the range was a range which one can access from selection.getRangeAt().  Selection should not reveal nodes in shadow dom.
Also it might be a bit surprising if selection is preserved in some cases but not in others.

Maybe, unrelated to moveBefore, there should be some helper methods to create a StaticRange from Range, and then update selection later, if/when needed, using that StaticRange. Hmm, though StaticRange isn't quite enough, since one needs also direction.

domfarolino replied:
OK the latest commit reverts special range handling, treating ranges just like we would in a normal remove situation. I'll resolve this.

annevk commented on dom.bs:
```diff
@@ -4986,6 +5151,16 @@ within <a>this</a>.
+
+<ol>
+ <li><p>Let <var>return node</var> be the result of <a>pre-moving</a> <var>node</var> into
+ <a>this</a> before <var>child</var>, rethrowing any exceptions.</p></li>
```
We don't say rethrow.

sorvell commented:
Would it be possible to have this method fallback to `insertBefore` if the conditions for an atomic move are not met? 

Without this, in practice it seems likely that users will need to write a boilerplate function that tests the 2 nodes for `isConnected` before calling. 

```js
const moveOrInsertNode = (container, node, ref_node = null) => {
 const canMove =  (container.isConnected && node.isConnected && (!ref_node || ref_node.parentNode === container);
  return canMove ? container.moveBefore(node, ref_node) : container.insertBefore(node, ref_node);
}
```

WebReflection commented:
FWIWI I fully agree with @sorvell ... the whole point of this proposal was to simplify DOM manipulation, not to complicate it even further with tons of repeated checks or `try/catch` around nodes that might not be fully owned by the underlying logic. The fallback to `insertBefore` when it fails is inevitable, as the intent is "*to move that node in that place*", nothing else, so it'd be counter-productive to have many checks around such intent + a try/catch toplay it safely plus the inevitable `insertBefore` on the catch.

domfarolino commented:
I'm personally pretty sympathetic to the arguments above. It was the original direction we went in, and I do think it is simpler to use, however the discussion we had at TPAC resulted in editors pushing back, on account of "move" being a fundamentally different primitive from "insert". @annevk are you open to revisiting this?

tjallingt commented:
"Move" would still be a fundamentally different primitive if it tried to preserve state where possible since the "insert" guarantees a default state. I understand that from a spec perspective this makes "move" a superset of "insert" given that it may need to run initialization hooks but I think that is a more usable primitive for developers.

domfarolino commented:
I definitely understand that. `moveBefore()` appears to be easier to use if it didn't throw any new errors that `insertBefore()`  doesn't already throw. At the same time, there is a predictability cost to this. As I [mentioned over in the Blink I2S](https://groups.google.com/a/chromium.org/g/blink-dev/c/YE_xLH6MkRs/m/ba-KZPUHAQAJ), it might  be weird if `moveBefore()` sometimes had huge side effects like running script, initializing iframes, applying styles, etc., yet other times no side effects at all (because the "atomic" move was carried out).

That said, if 99% of `moveBefore()` uses are from a generic place like a framework, where you _always_ want to move the node from A->B (just atomically when possible!) then the error-throwing behavior becomes cumbersome, because then 99% of users have to call `insertBefore()` from a catch block just to use the API correctly.

I guess it comes down to how often we think people actually care to know whether a truly "atomic" move can be performed, and might act differently if they find out it can't. That's where the error-throwing behavior becomes really useful; without it, if you want to know whether `moveBefore()` will actually move atomically, you have to remember to check [all of the pre-conditions](https://whatpr.org/dom/1307.html#concept-node-ensure-pre-move-validity) before calling `moveBefore()`.

noamr commented:
I think that the following scenario is plausible:

```js
try {
  element.moveBefore(newNode, refNode);
} catch {
  element.insertBefore(newNode, refNode);
}  finally {
  // check whether we need to do something extra - e.g. the focused element reparented
  // and some library relies on the side effects
}
```

It's likely that the `finally` clause would be empty in the vast majority of cases, however it's debatable whether it's the platform's job to hide this difference for the sake of ergonomics in a somewhat low-level API like DOM.

dead-claudia commented:
I wonder if that choice of error tolerance could be moved to a boolean argument. Like, `parent.moveBefore(child, refNode, true)` to tolerate failure to move (like a missing custom element move callback) and `parent.moveBefore(child, refNode, false)` to throw on failure, or vice versa. That could be an acceptable compromise.

Also, how does this handle the case of `parent` and `child` having different `Document`s? If I'm understanding correctly, an exception is thrown, but that's just a guess that checking "shadow-including root" equality implies that.

noamr commented:
> > however it's debatable whether it's the platform's job to hide this difference for the sake of ergonomics in a somewhat low-level API like DOM.
> 
> @noamr There _is_ some precedent for light ergonomics in the DOM already: `child.before`, `child.after`, `parent.append`, and `parent.replaceChildren` all four accepting strings and auto-converting them into text nodes. I feel this could qualify as such "light ergonomics".

Yes, and we can certainly consider adding more ergonomic variants in the future on top of the `moveBefore` behavior, the same way `appendChild` is an ergonomic version of `insertBefore`.
However, `moveBefore` itself, being the first primitive to support the "move" operation, is not that - there is value in making it as single-purpose as possible: it only moves. It doesn't fall back to something else, but fails and notifies the author.

noamr commented:
> I wonder if that choice of error tolerance could be moved to a boolean argument. Like, `parent.moveBefore(child, refNode, true)` to tolerate failure to move (like a missing custom element move callback) and `parent.moveBefore(child, refNode, false)` to throw on failure, or vice versa. That could be an acceptable compromise.

This particular API shape is a boolean trap, but I think we can can consider variants in the future that have an 'move if you can insert if you can't' behavior. But this is a slightly higher level behavior than moving. As I said, there is value in having at least some DOM APIs that are as primitive as possible.

> Also, how does this handle the case of `parent` and `child` having different `Document`s? If I'm understanding correctly, an exception is thrown, but that's just a guess that checking "shadow-including root" equality implies that.

It throws when moving across documents.

WebReflection commented:
> it might be weird if moveBefore() sometimes had huge side effects like running script, initializing iframes, applying styles, etc., yet other times no side effects at all (because the "atomic" move was carried out).

if everyone will inevitably end up using a `try/catch` that fallback to `insertBefore` this API throwing doesn't help at all + all checks [mentioned in here](https://github.com/whatwg/dom/pull/1307#issuecomment-2491906616) are not obvious, developer friendly, or performance friendly.

When a developer does not want to have side-effects it can perform those checks manually, the rest 99% of the world will just `moveBefore` and fallback to `insertBefore`, including all frameworks, hence my question: who is this throwing behavior helping and why an alternative solution such as `parent.canMoveNode(node)` has not been discussed, as that would do surely the right thing internally as opposite of asking developers to know what are all conditions to check before a node can be moved or not?

Please note I am not suggesting `canMoveNode` together with a throwing `moveBefore`, I am suggesting a `moveBefore` that tries to do the right thing and fallback to `insertBefore` without throwing, plus a `canMoveNode` for anyone interested in that detail (which is niece, or tests based things, or perf based things, or debugging based things).

In short, if try/catch is the workaround, everyone will use it and all the reasons it couldn't be done behind the scene in the name of "*it's better for developers*" will be futile, I hope we can agree on that.

annevk commented:
The reason it throws is because in the future we might become more ambitious and also tackle those scenarios. The reason we don't have various accompanying APIs such as `canMoveBefore()` is because we tend to develop APIs in baby steps. Doing it all at once typically leads to regrets. Please continue this discussion in #1255 as otherwise this PR becomes unmanageable.

annevk reviewed (changes requested):
The aspects shared between insert and move, and remove and move, need to be shared. We don't want to have to maintain identical range mutation in multiple places.

annevk commented on dom.bs:
```diff
@@ -2652,6 +2652,63 @@ of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run the
+ <var>node</var>'s <a for=/>shadow-including root</a>, then <a>throw</a> a
+ "{{HierarchyRequestError!!exception}}" {{DOMException}}.</p></li>
+
+ <li><p>If <var>parent</var> is not an {{Element}} or {{DocumentFragment}} <a for=/>node</a>, then
+ <a>throw</a> a "{{HierarchyRequestError!!exception}}" {{DOMException}}.
```
The reason this restriction is okay is because documents cannot have multiple element children and you can only use `insertBefore()` at the root when the current element is removed. So it really relies on the connected checks above. That's very subtle.

However, when looking at non-elements it's quite a bit more restrictive than what it probably should be. I don't see why we'd forbid moving a comment to be a child of a document. So I would suggest we do this differently, even though in practice there's no benefit to moving comments and the like.

Ideally we'd align this with "pre-insertion validity" or share as much as possible.

smaug---- replied:
Yeah, I'd expect this to handle basically ParentNode mixin.

domfarolino replied:
Hmm, OK. Would you prefer we simply make this condition perfectly match the pre-insertion validity steps? I.e., expanding it to include Document parents too? I'm fine with that. Let me know what you think @annevk.

> Yeah, I'd expect this to handle basically ParentNode mixin.

Hmm, I'm not exactly sure what this means. But does what I recommend above satisfy what you're looking for?

annevk replied:
Yeah, I think that would be good, although I have not thought through all implications of such a change.

domfarolino replied:
OK I've done that in https://github.com/whatwg/dom/pull/1307/commits/90c9a943a859267acd05b37c48120427e497c7e2, and changed the Chromium implementation and added a test in https://crrev.com/c/6081493. I agree this is a pretty minor change, especially since Documents can still only have a single Element child, and it is nice to have consistency with the pre-insertion validity conditions. I'll resolve this now.

annevk commented on dom.bs:
```diff
@@ -2845,6 +2902,124 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+   <a>enqueue a custom element callback reaction</a> with <var>inclusiveDescendant</var>, callback
+   name "<code>connectedMoveCallback</code>", and an empty argument list.
+
+   <li>
+    <p>Otherwise, <a lt="try to upgrade an element">try to upgrade</a>
+    <var>inclusiveDescendant</var>.
+
+    <p class=note>If this successfully upgrades <var>inclusiveDescendant</var>, its
+    <code>connectedCallback</code> will be enqueued automatically during the
+    <a>upgrade an element</a> algorithm.
+   </li>
```
How could this happen due to a move? That seems incorrect?

noamr replied:
I guess if we had scoped element registries and we moved across trees?

domfarolino replied:
OK yeah I had to get educated a bit more on exactly how custom elements work and upgrade, but I agree this is not relevant for this PR. It's now removed; resolving.

annevk commented on dom.bs:
```diff
@@ -3977,8 +4152,8 @@ method steps are:
  <li><p><a>Queue a mutation record</a> of "<code>childList</code>" for <var>target</var> with
- null, null, null, <var>addedNodes</var>, <var>removedNodes</var>, <var>previousSibling</var>,
- and <var>nextSibling</var>.
+ null, null, null, <var>addedNodes</var>, <var>removedNodes</var>, <var>previousSibling</var>, and
+ <var>nextSibling</var>.
```
Let's not change this in this PR.

annevk commented on dom.bs:
```diff
@@ -4118,6 +4293,8 @@ interface Node : EventTarget {
   [CEReactions] Node appendChild(Node node);
   [CEReactions] Node replaceChild(Node node, Node child);
   [CEReactions] Node removeChild(Node child);
+
```
I don't think we need this newline here.

smaug---- commented on dom.bs:
```diff
@@ -2652,6 +2652,63 @@ of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run the
+   <li><p><var>parent</var> is <a>connected</a> and <var>node</var> is not <a>connected</a>; or</p></li>
+
+   <li><p><var>parent</var> is not <a>connected</a> and <var>node</var> is <a>connected</a>,</p></li>
+  </ul>
```
Why do we need this when we have " If parent ’s shadow-including root is not the same as node ’s shadow-including root " check below?

domfarolino replied:
So, I think back when we only supported the connected->connected case, the "shadow-including root matches" check was right, because "shadow-including root" was essentially the same as "node document", in the connected case.

But now that we support the disconnected->disconnected case, I actually think the "shadow-including root" check is too strict, right? For example, imagine:

```js
const divA = document.createElement('div');
const p = divA.appendChild(document.createElement('p'));
const divB = document.createElement('div');

// Throws because divB's shadow-including root does not match p's, since they're both disconnected.
divB.moveBefore(p, null);
```

That is, in the disconnected->disconnected case, it is way more likely that the target node and destination container's shadow-including roots will not match, since in the disconnected case, a shadow-including root just tops-out at the nearest disconnected parent. So my proposal is to simply change the shadow-including root comparison to instead compare node documents. I'll do that now, and please let me know if you think that sounds good.

smaug---- replied:
Well, I don't want to support move between otherwise unrelated subtrees. This all becomes simpler if we don't have that. No need to check same document or anything like that.

domfarolino replied:
> Well, I don't want to support move between otherwise unrelated subtrees.

What are "otherwise unrelated subtrees"? We certainly want to support the code example I wrote above, no?

annevk replied:
I think Olli always wants there to be a shared ancestor.

domfarolino replied:
Yeah, I personally don’t think I see the value in enforcing such a requirement. Do you feel strongly about this? I think folks should be able to create detached DocumentFragments and move things around from them to possibly other detached subtrees, and mostly I don’t see the value of enforcing a common ancestor between old and new parent. What do you think @annevk ?

annevk replied:
I'm not sure I hold a strong opinion, but I do like the symmetry Olli is suggesting.

domfarolino replied:
Hmm, I'm not sure it provides any more symmetry than the same-document check being enforced in the connected and disconnected case. But yeah I'll just replace this check with the shadow-inclusive root check to unblock this PR, and we'll update the tests.

domfarolino replied:
Done in https://github.com/whatwg/dom/pull/1307/commits/65b4e13ade1039eee8e49316b3728e9f4191c414.

domfarolino commented:
All comments appear to be addressed at this point. PTAL!

annevk commented:
I've reopened the validity topic. I think you now allow elements to be moved into document's children without any kind of additional checks. The checks are also in a different order. And given how many are shared it probably warrants more abstraction. (Also, "pre-insertion", but "pre-move" is weird. Maybe "pre-insertion" is wrong?)

domfarolino commented:
> I think you now allow elements to be moved into document's children without any kind of additional checks.

OK, I've fixed this, adding the relevant pre-insertion conditions to the pre-move algorithm. I'll do the same for Chromium and WPTs.

> The checks are also in a different order.

Fixed. The pre-move-specific checks still come first, but all **relevant** pre-insert conditions come after, and in the same order as the appear in the pre-insertion algorithm.

> And given how many are shared it probably warrants more abstraction.

Hmm, I could go either way. There are not that many conditions in pre-move that are carbon copies of their corresponding conditions pre-insert conditions. Perhaps the first 3 can be cleanly lifted out now, but the remaining are modified enough to probably not warrant abstraction. Would you prefer I lift the first 3 conditions out into a common sub algorithm?

> (Also, "pre-insertion", but "pre-move" is weird. Maybe "pre-insertion" is wrong?)

I've renamed all "pre-insertion" instances to "pre-insert".

annevk commented on dom.bs:
```diff
@@ -2845,6 +2911,101 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+
+   <li><p>If <var>inclusiveDescendant</var> is <a for=Element>custom</a> and <var>newParent</var> is
+   <a>connected</a>, then <a>enqueue a custom element callback reaction</a> with
+   <var>inclusiveDescendant</var>, callback name "<code>connectedMoveCallback</code>", and an empty
+   argument list.
```
```suggestion
   <var>inclusiveDescendant</var>, callback name "<code>connectedMoveCallback</code>", and « ».
```
We also have to clean this up elsewhere, but that can be done separately. New algorithms should do it right from the start though.

annevk commented on dom.bs:
```diff
@@ -8163,6 +8320,31 @@ but not its <a for=range>end node</a>, or vice versa.
  </ul>
 </div>
 
+<p>The <dfn>live range pre-removing steps</dfn> given a <a>node</a> <var>node</var>, are as follows:
```
Should this not be pre-remove steps?

annevk commented on dom.bs:
```diff
@@ -4118,6 +4264,7 @@ interface Node : EventTarget {
   [CEReactions] Node appendChild(Node node);
   [CEReactions] Node replaceChild(Node node, Node child);
   [CEReactions] Node removeChild(Node child);
+  [CEReactions] Node moveBefore(Node node, Node? child);
```
I think we should put this method on `ParentNode`. That would also allow us to simplify the validity checks a bit.

annevk replied:
It also needs a "web developer" intro box.

domfarolino replied:
Done in https://github.com/whatwg/dom/pull/1307/commits/3f3e94c5beda922962dacaeb606087f57bd7f7be! Besides the exposure of the method, the semantics of that commit are not observable, right?

domfarolino commented on dom.bs:
```diff
@@ -2845,6 +2911,101 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+
+
+<p>To <dfn export id=concept-node-move>move</dfn> a <var>node</var> into a <var>newParent</var>
+before a <var>child</var>, run these steps:
```
This was inspired by https://dom.spec.whatwg.org/#concept-node-insert, https://dom.spec.whatwg.org/#concept-node-ensure-pre-insertion-validity, and https://dom.spec.whatwg.org/#concept-node-pre-insert. Should I still remove it?

annevk replied:
Yes.

domfarolino commented on dom.bs:
```diff
@@ -2652,6 +2652,72 @@ of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run the
+ <li><p>If <var>node</var> is a {{Text}} <a for=/>node</a> and <var>parent</var> is a
+ <a>document</a>, then <a>throw</a> a "{{HierarchyRequestError!!exception}}" {{DOMException}}.
+
+ <li><p>If <var>parent</var> is a {{Document}}, <var>node</var> is an {{Element}}
```
Out of curiosity, why this change? Earlier in this algorithm we do a very similar comparison of `<var>parent</var>`, but with `{{Document}}`.

annevk replied:
I think in that case there's multiple nodes so it's clearer to go for the longer "X node" concept as there's multiple of X. You could do "Document node" here too, but generally we shorten that where we can.

peasantman1 commented:
> This PR introduces a new DOM API on the `Node` interface: `moveBefore()`. It mirrors `insertBefore()` in shape, but defers to a new DOM manipulation primitive that this PR adds in service of this new API: the "move" primitive. The move primitive contains some of the DOM tree bookkeeping steps from the remove primitive, as well as the insert primitive, and does three more interesting things:
> 
> 1. Calls the [moving steps hook](https://whatpr.org/dom/1307.html#concept-moving-steps-ext) with the moved node and the old parent (possibly null, just like the removing steps)
> 2. Queues a custom element callback reaction for the `connectedMoveCallback()`
> 3. Queues two back-to-back mutation record tasks: one for removal from the old parent; one for insertion into the new one
> 
> The power of the move primitive comes from the fact that the algorithm does not defer to the traditional insert and removal primitives, and therefore does not invoke the [removing steps](https://dom.spec.whatwg.org/#concept-node-remove-ext) and [insertion steps](https://dom.spec.whatwg.org/#concept-node-insert-ext). This allows most state to be preserved by default (i.e., we don't tear down iframes, or close dialogs). Sometimes, the insertion/removing step overrides in other specifications have steps that do need to be performed during a move anyways. These specifications are expected to override the moving steps hook and perform the necessary work accordingly. See [whatwg/html#10657](https://github.com/whatwg/html/pull/10657) for HTML.
> 
> Remaining tasks (some will be PRs in other standards):
> 
> * [x]  Custom element integration
> * [x]  Keep popovers open
> * [x]  Don't call post-connection steps if state-preserving atomic move is in progress
> * [x]  Don't call becomes connected / becomes browsing-context
> * [x]  Only disconnect subframes on removal when state-preserving atomic move is **not** in progress
> * [x]  Keep dialogs open: see [removing steps](https://html.spec.whatwg.org/C#the-dialog-element:html-element-removing-steps)
> * [x]  img/source: this shouldn't count as a [relevant mutation](https://html.spec.whatwg.org/C#relevant-mutations)
> * [x]  Preserve fullscreen
> * [x]  Preserve focus 
>   
>   * Need to resolve `focusin` event semantics
> * ~[ ] Don't reset animations / transitions. See [here](https://drafts.csswg.org/css-transitions/#starting) 
>   
>   * Maybe nothing needs to be done here. Given [how element removals are handled](https://drafts.csswg.org/css-transitions/#transition-cancel:~:text=If%20an%20element%20is%20no%20longer%20in%20the%20document%2C%20implementations%20must%20cancel%20any%20running%20transitions%20on%20it%20and%20remove%20transitions%20on%20it%20from%20the%20completed%20transitions), the spec does NOT require transitions to be removed from the UA's set of _running transitions_ for moved nodes since they are never removed from the Document.~
> * ~[ ] Preserve text-selection. See [set the selection range](https://html.spec.whatwg.org/C#set-the-selection-range). Edit: Nothing needs to be done here. Selection metadata (i.e., `selectionStart` and kin) is preserved by default in browsers, consistent with HTML (no action is taken on removal). The UI behavior of the selection not being highlighted is a side-effect of the element losing focus~
> * [ ]  Selection API: don't reset the Document's [selection](https://w3c.github.io/selection-api/#dfn-selection)
>   
>   * Updates to the selection range should happen according to how the DOM Standard primitives update ranges. The Selection API specification admits as much, by [deferring to the](https://w3c.github.io/selection-api/#responding-to-dom-mutations) insert and removal algorithms. Therefore, we should reference the move primitive from the Selection API specification, and ensure that the move primitive in _this DOM Standard PR_ updates live ranges correctly: [Reference the move primitive in DOM mutations section w3c/selection-api#341](https://github.com/w3c/selection-api/pull/341)
>   * `selectionchange` event: We've decided to allow `selectionchange` event to still fire, [since it is queued in a task](https://w3c.github.io/selection-api/#ref-for-dfn-schedule-a-selectionchange-event-1). No changes for this part are required.
> * [x]  Pointer event state reset: see [here](https://w3c.github.io/pointerevents/#suppressing-a-pointer-event-stream)
>   
>   * The specification does not reset an element's previously-set pointer capture upon DOM node insertion or removal. While this seems like a gap in that specification ([[DOM integration] No state is reset on node removal w3c/pointerevents#526](https://github.com/w3c/pointerevents/issues/526)), it appears that nothing needs to be done in that specification to satisfy this preservation semantic.
>   * Test: https://github.com/web-platform-tests/wpt/blob/master/dom/nodes/moveBefore/tentative/pointer-events.html
> * [x]  Hide input/select picker: [here](https://html.spec.whatwg.org/C#show-the-picker%2C-if-applicable)
>   
>   * Nothing in the specification seems to hook into DOM for removal when the picker dialog is shown, so no work needs to be done in this PR. See ["Show the picker" steps never dismisses the picker upon node removal html#10743](https://github.com/whatwg/html/issues/10743) for a follow-up.
> * [x]  Preserve pointer lock: [here](https://w3c.github.io/pointerlock/#requestPointerLock)
>   
>   * When a ["pointer-lock target becomes disconnected"](https://w3c.github.io/pointerlock/#ref-for-dfn-pointer-lock-target-12), pointer lock is exited. Since the "becomes disconnected" hook is not run during an atomic move, pointer lock is preserved naturally.
> * [ ]  Containment: keep last remembered size(see [here](https://drafts.csswg.org/css-sizing-4/#last-remembered))
> 
> * [x]  At least two implementers are interested (and none opposed):
>   
>   * Chromium: https://issues.chromium.org/issues/40150299
>   * WebKit: [DOM State-preserving move (`Node.prototype.moveBefore`) WebKit/standards-positions#375](https://github.com/WebKit/standards-positions/issues/375)
>   * Gecko: [DOM State-preserving move (`Node.prototype.moveBefore`) mozilla/standards-positions#1053](https://github.com/mozilla/standards-positions/issues/1053)
> * [x]  [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
>   
>   * https://github.com/web-platform-tests/wpt/tree/master/dom/nodes/moveBefore/tentative (need to mark as non-tentative)
> * [x]  [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
>   
>   * Chromium: https://issues.chromium.org/issues/40150299
>   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1923880
>   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=281223
> * [x]  [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: I've added the impacts-documentation label, per https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests
> * [x]  The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use.
> 
> (See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)
> 
> [Preview](https://whatpr.org/dom/1307.html) | [Diff](https://whatpr.org/dom/1307/719e45b...4b54a0d.html)

 Help

peasantman1 commented:
> This PR introduces a new DOM API on the `Node` interface: `moveBefore()`. It mirrors `insertBefore()` in shape, but defers to a new DOM manipulation primitive that this PR adds in service of this new API: the "move" primitive. The move primitive contains some of the DOM tree bookkeeping steps from the remove primitive, as well as the insert primitive, and does three more interesting things:
> 
> 1. Calls the [moving steps hook](https://whatpr.org/dom/1307.html#concept-moving-steps-ext) with the moved node and the old parent (possibly null, just like the removing steps)
> 2. Queues a custom element callback reaction for the `connectedMoveCallback()`
> 3. Queues two back-to-back mutation record tasks: one for removal from the old parent; one for insertion into the new one
> 
> The power of the move primitive comes from the fact that the algorithm does not defer to the traditional insert and removal primitives, and therefore does not invoke the [removing steps](https://dom.spec.whatwg.org/#concept-node-remove-ext) and [insertion steps](https://dom.spec.whatwg.org/#concept-node-insert-ext). This allows most state to be preserved by default (i.e., we don't tear down iframes, or close dialogs). Sometimes, the insertion/removing step overrides in other specifications have steps that do need to be performed during a move anyways. These specifications are expected to override the moving steps hook and perform the necessary work accordingly. See [whatwg/html#10657](https://github.com/whatwg/html/pull/10657) for HTML.
> 
> Remaining tasks (some will be PRs in other standards):
> 
> * [x]  Custom element integration
> * [x]  Keep popovers open
> * [x]  Don't call post-connection steps if state-preserving atomic move is in progress
> * [x]  Don't call becomes connected / becomes browsing-context
> * [x]  Only disconnect subframes on removal when state-preserving atomic move is **not** in progress
> * [x]  Keep dialogs open: see [removing steps](https://html.spec.whatwg.org/C#the-dialog-element:html-element-removing-steps)
> * [x]  img/source: this shouldn't count as a [relevant mutation](https://html.spec.whatwg.org/C#relevant-mutations)
> * [x]  Preserve fullscreen
> * [x]  Preserve focus 
>   
>   * Need to resolve `focusin` event semantics
> * ~[ ] Don't reset animations / transitions. See [here](https://drafts.csswg.org/css-transitions/#starting) 
>   
>   * Maybe nothing needs to be done here. Given [how element removals are handled](https://drafts.csswg.org/css-transitions/#transition-cancel:~:text=If%20an%20element%20is%20no%20longer%20in%20the%20document%2C%20implementations%20must%20cancel%20any%20running%20transitions%20on%20it%20and%20remove%20transitions%20on%20it%20from%20the%20completed%20transitions), the spec does NOT require transitions to be removed from the UA's set of _running transitions_ for moved nodes since they are never removed from the Document.~
> * ~[ ] Preserve text-selection. See [set the selection range](https://html.spec.whatwg.org/C#set-the-selection-range). Edit: Nothing needs to be done here. Selection metadata (i.e., `selectionStart` and kin) is preserved by default in browsers, consistent with HTML (no action is taken on removal). The UI behavior of the selection not being highlighted is a side-effect of the element losing focus~
> * [ ]  Selection API: don't reset the Document's [selection](https://w3c.github.io/selection-api/#dfn-selection)
>   
>   * Updates to the selection range should happen according to how the DOM Standard primitives update ranges. The Selection API specification admits as much, by [deferring to the](https://w3c.github.io/selection-api/#responding-to-dom-mutations) insert and removal algorithms. Therefore, we should reference the move primitive from the Selection API specification, and ensure that the move primitive in _this DOM Standard PR_ updates live ranges correctly: [Reference the move primitive in DOM mutations section w3c/selection-api#341](https://github.com/w3c/selection-api/pull/341)
>   * `selectionchange` event: We've decided to allow `selectionchange` event to still fire, [since it is queued in a task](https://w3c.github.io/selection-api/#ref-for-dfn-schedule-a-selectionchange-event-1). No changes for this part are required.
> * [x]  Pointer event state reset: see [here](https://w3c.github.io/pointerevents/#suppressing-a-pointer-event-stream)
>   
>   * The specification does not reset an element's previously-set pointer capture upon DOM node insertion or removal. While this seems like a gap in that specification ([[DOM integration] No state is reset on node removal w3c/pointerevents#526](https://github.com/w3c/pointerevents/issues/526)), it appears that nothing needs to be done in that specification to satisfy this preservation semantic.
>   * Test: https://github.com/web-platform-tests/wpt/blob/master/dom/nodes/moveBefore/tentative/pointer-events.html
> * [x]  Hide input/select picker: [here](https://html.spec.whatwg.org/C#show-the-picker%2C-if-applicable)
>   
>   * Nothing in the specification seems to hook into DOM for removal when the picker dialog is shown, so no work needs to be done in this PR. See ["Show the picker" steps never dismisses the picker upon node removal html#10743](https://github.com/whatwg/html/issues/10743) for a follow-up.
> * [x]  Preserve pointer lock: [here](https://w3c.github.io/pointerlock/#requestPointerLock)
>   
>   * When a ["pointer-lock target becomes disconnected"](https://w3c.github.io/pointerlock/#ref-for-dfn-pointer-lock-target-12), pointer lock is exited. Since the "becomes disconnected" hook is not run during an atomic move, pointer lock is preserved naturally.
> * [ ]  Containment: keep last remembered size(see [here](https://drafts.csswg.org/css-sizing-4/#last-remembered))
> 
> * [x]  At least two implementers are interested (and none opposed):
>   
>   * Chromium: https://issues.chromium.org/issues/40150299
>   * WebKit: [DOM State-preserving move (`Node.prototype.moveBefore`) WebKit/standards-positions#375](https://github.com/WebKit/standards-positions/issues/375)
>   * Gecko: [DOM State-preserving move (`Node.prototype.moveBefore`) mozilla/standards-positions#1053](https://github.com/mozilla/standards-positions/issues/1053)
> * [x]  [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
>   
>   * https://github.com/web-platform-tests/wpt/tree/master/dom/nodes/moveBefore/tentative (need to mark as non-tentative)
> * [x]  [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
>   
>   * Chromium: https://issues.chromium.org/issues/40150299
>   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1923880
>   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=281223
> * [x]  [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: I've added the impacts-documentation label, per https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests
> * [x]  The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use.
> 
> (See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)
> 
> [Preview](https://whatpr.org/dom/1307.html) | [Diff](https://whatpr.org/dom/1307/719e45b...4b54a0d.html)

Help

smaug---- commented on dom.bs:
```diff
@@ -2872,6 +2872,128 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+ <var>newParent</var> is a <a>slot</a> whose <a for=slot>assigned nodes</a> is the empty list,
+ then run <a>signal a slot change</a> for <var>newParent</var>.
+
+ <li><p>Run <a>assign slottables for a tree</a> with <var>node</var>'s <a for=tree>root</a>.
```
Does this happen somewhere also for the node's old parent? We need to re-assign slots also there.
(and are there tests for this?)

smaug---- replied:
I didn't immediately find any tests for moveBefore + slotchange event

domfarolino replied:
OK I've added tests [here](https://chromium-review.googlesource.com/c/chromium/src/+/6111212) (pending) and addressed this in the spec with: https://github.com/whatwg/dom/pull/1307/commits/32a8a683e49c2772d08a622413500f69d0455f3a.

smaug---- commented:
Are there any tests for MutationObserver + moveBefore?

domfarolino commented:
> Are there any tests for MutationObserver + moveBefore?

I thought there were but I cannot find them, so I'll add some today.

domfarolino commented:
OK `slotchange` and `MutationObserver` tests are added in https://github.com/web-platform-tests/wpt/pull/49810 and https://chromium-review.googlesource.com/c/chromium/src/+/6108064 (upstreamed soon), so I think everything here is resolved. I'll look at the comments on the HTML integration PR, but I think this one is ready to go.

annevk commented on dom.bs:
```diff
@@ -2872,6 +2872,135 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
 
 
+<p><a lt="Other applicable specifications">Specifications</a> may define <dfn export
+id=concept-node-move-ext>moving steps</dfn> for all or some <a for=/>nodes</a>. The algorithm is
```
The DOM standard doesn't wrap on all spaces. Phrasing-level elements need to be kept on a single line. (This applies several times in this PR.)

It also seems okay to drop the explicit ID here as we stopped doing those.

annevk commented on dom.bs:
```diff
@@ -2872,6 +2872,135 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+id=concept-node-move-ext>moving steps</dfn> for all or some <a for=/>nodes</a>. The algorithm is
+passed a <a for=/>node</a> <var ignore>movedNode</var>, and a <a for=/>node</a>-or-null <var
+ignore>oldParent</var> as indicated in the <a for=/>move</a> algorithm below. Like the <span
+data-x="concept-insertion-steps-ext">insertion steps</span>, these steps must not modify the
```
This is not how cross-references work in the DOM standard. There's no `data-x`. I'm surprised this passes CI.

annevk reviewed (commented):
I can fix these nits before merging if you want. I will merge this and the HTML PR on January 20 provided there is no further feedback.

smaug---- commented on dom.bs:
```diff
@@ -2872,6 +2872,135 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+
+ <li><p>If <var>oldParent</var>'s <a for=tree>root</a> is a <a for=/>shadow root</a>, and
+ <var>oldParent</var> is a <a>slot</a> whose <a for=slot>assigned nodes</a> is the empty list, then
+ run <a>signal a slot change</a> for <var>oldParent</var>.
```
Aren't we missing the steps needed to track the case when slot element itself if moving. In the "remove" algorithm step 14.
And I didn't see tests for moving slots.

domfarolino replied:
Apologies for the delay on this. Yeah I think you're right. I've added the missing step in https://github.com/whatwg/dom/pull/1307/commits/87b6acbf12368302a05b22ed349d7bacd89b1507 and a test for this in https://chromium-review.googlesource.com/c/chromium/src/+/6298941 which will be upstreamed soon.

annevk commented on dom.bs:
```diff
@@ -2881,6 +2881,145 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+algorithm is passed a <a for=/>node</a> <var ignore>movedNode</var>, and a <a for=/>node</a>-or-null
+<var ignore>oldParent</var> as indicated in the <a for=/>move</a> algorithm below. Like the
+<a>insertion steps</a>, these steps must not modify the <a>node tree</a> that
+<var>insertedNode</var> <a>participates</a> in, create <a for=/>browsing contexts</a>,
```
```suggestion
<var>movedNode</var> <a>participates</a> in, create <a for=/>browsing contexts</a>,
```

annevk commented on dom.bs:
```diff
@@ -2881,6 +2881,145 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+ <var>node</var>'s <a>assigned slot</a>.
+
+ <li><p>If <var>oldParent</var>'s <a for=tree>root</a> is a <a for=/>shadow root</a>, and
+ <var>oldParent</var> is a <a>slot</a> whose <a for=slot>assigned nodes</a> is the empty list, then
```
```suggestion
 <var>oldParent</var> is a <a>slot</a> whose <a for=slot>assigned nodes</a> <a for=list>is empty</a>, then
```
(or something like that)

annevk commented on dom.bs:
```diff
@@ -2881,6 +2881,145 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+
+   <li><p>Run <a>assign slottables for a tree</a> with <var>node</var>.
+  </ol>
+ </li>
```
```suggestion
```

annevk commented on dom.bs:
```diff
@@ -2881,6 +2881,145 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+ <a>slottable</a>, then <a>assign a slot</a> for <var>node</var>.
+
+ <li><p>If <var>newParent</var>'s <a for=tree>root</a> is a <a for=/>shadow root</a>, and
+ <var>newParent</var> is a <a>slot</a> whose <a for=slot>assigned nodes</a> is the empty list,
```
empty list thing again

annevk commented on dom.bs:
```diff
@@ -3234,6 +3360,16 @@ Element includes ParentNode;
+ <dd>
+  <p>Moves, without first removing, <var>movedNode</var> into <var>node</var> after <var>child</var>
+  if <var>child</var> is non-null; otherwise after the <a>last child</a> of <var>node</var>. This
+  methods preserves state associated with <var>movedNode</var> so that it persists after the move.
```
```suggestion
  method preserves state associated with <var>movedNode</var>.
```
persists seems redundant with preserves

annevk commented on dom.bs:
```diff
@@ -3234,6 +3360,16 @@ Element includes ParentNode;
+
+  <p><a>Throws</a> a "{{HierarchyRequestError!!exception}}" {{DOMException}} if the constraints of
+  the <a>node tree</a> are violated.
+  <!-- "NotFoundError" is impossible -->
```
We should probably call out that we also throw this when we cannot preserve state.

domfarolino commented on dom.bs:
```diff
@@ -2881,6 +2881,145 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
+
+
+<p>To <dfn>move</dfn> a <a for=/>node</a> <var>node</var> into a <a for=/>node</a>
+<var>newParent</var> before a <a for=/>node</a> <var>child</var>:
```
That's right, fixing. Thanks.

annevk reviewed (approved):
@smaug---- can you double check as well? Is Monday a reasonable merge day? I feel like everyone else already got sufficient time in the last couple of rounds.

smaug---- commented:
I think this is fine.

https://issues.chromium.org/issues/400758510 is interesting. That was filed immediately when someone tried to use the API. That is more for the HTML part of the feature though. But I think opt-in behavior of the new callback is reasonable (though the callback should likely get old parent as a param, but that can be a followup)

peasantman1 commented:
These people are using federated learning to  download information in my device
